File: scripts/generate_latin_squares.py
import random
import uuid
import json
import os
import hashlib
import copy

# Ensure migration directory exists
os.makedirs("supabase/migrations", exist_ok=True)
output_file = "supabase/migrations/20260723000006_batch2_latin_squares.sql"

def generate_full_latin_square():
    # Base 5x5
    base = [
        ['A', 'B', 'C', 'D', 'E'],
        ['B', 'C', 'D', 'E', 'A'],
        ['C', 'D', 'E', 'A', 'B'],
        ['D', 'E', 'A', 'B', 'C'],
        ['E', 'A', 'B', 'C', 'D']
    ]
    # Shuffle rows
    random.shuffle(base)
    # Shuffle columns
    base = [list(x) for x in zip(*base)]
    random.shuffle(base)
    base = [list(x) for x in zip(*base)]
    # Shuffle symbols
    symbols = ['A', 'B', 'C', 'D', 'E']
    shuffled_symbols = symbols.copy()
    random.shuffle(shuffled_symbols)
    mapping = {symbols[i]: shuffled_symbols[i] for i in range(5)}
    
    square = []
    for r in range(5):
        row = []
        for c in range(5):
            row.append(mapping[base[r][c]])
        square.append(row)
    return square

def is_valid(grid, r, c, val):
    for i in range(5):
        if grid[r][i] == val or grid[i][c] == val:
            return False
    return True

def count_solutions_for_target(grid, target_r, target_c, target_val):
    # We want to know if the target cell is forced to be target_val.
    # To do this, we try all OTHER possible values for the target cell.
    # If ANY other value leads to a valid completely filled grid (or just a valid partial solution that doesn't violate rules), 
    # then the target is not uniquely determined. 
    # Actually, a full solver to check uniqueness of the ENTIRE grid is safer.
    solutions = []
    
    def solve(g, r, c):
        if len(solutions) > 1: return
        if r == 5:
            solutions.append(copy.deepcopy(g))
            return
        if c == 5:
            solve(g, r+1, 0)
            return
        if g[r][c] != '':
            solve(g, r, c+1)
            return
            
        for val in ['A', 'B', 'C', 'D', 'E']:
            if is_valid(g, r, c, val):
                g[r][c] = val
                solve(g, r, c+1)
                g[r][c] = ''
                
    g_copy = copy.deepcopy(grid)
    solve(g_copy, 0, 0)
    return len(solutions)

def create_puzzle(difficulty):
    while True:
        full = generate_full_latin_square()
        target_r, target_c = random.randint(0, 4), random.randint(0, 4)
        target_val = full[target_r][target_c]
        
        # Start with all cells filled
        puzzle = copy.deepcopy(full)
        puzzle[target_r][target_c] = '?'
        
        # List of other cells
        cells = [(r, c) for r in range(5) for c in range(5) if (r, c) != (target_r, target_c)]
        random.shuffle(cells)
        
        removed_count = 0
        target_removals = 12 if difficulty == 'easy' else (15 if difficulty == 'medium' else 18)
        
        for r, c in cells:
            if removed_count >= target_removals:
                break
            temp = puzzle[r][c]
            puzzle[r][c] = ''
            
            # Check if puzzle still has exactly one solution
            if count_solutions_for_target(puzzle, target_r, target_c, target_val) == 1:
                removed_count += 1
            else:
                puzzle[r][c] = temp
                
        if removed_count >= target_removals - 2: # accept slightly fewer removals if it's hard to find
            return puzzle, target_val

def format_grid(grid):
    lines = []
    for row in grid:
        # Pad cells to align nicely
        formatted_row = [cell if cell != '' else '_' for cell in row]
        lines.append(" | ".join(formatted_row))
    return "\n".join(lines)

def main():
    num_questions = 100
    
    sql_statements = []
    sql_statements.append("-- ==============================================================================")
    sql_statements.append("-- Migration: Batch 2 - 100 Programmatically Generated Latin Squares")
    sql_statements.append("-- ==============================================================================\n")
    sql_statements.append("INSERT INTO public.questions (id, module, prompt_text, options, correct_option_id, explanation_text, difficulty, source_id, time_estimate_seconds, subtopic, question_type, content_hash) VALUES")
    
    values = []
    difficulties = ['easy', 'medium', 'hard']
    
    for i in range(num_questions):
        diff = random.choice(difficulties)
        puzzle, target_val = create_puzzle(diff)
        
        grid_str = format_grid(puzzle)
        prompt = f"In the following 5x5 Latin square, each of the letters A, B, C, D, and E must appear exactly once in each row and each column.\n\n{grid_str}\n\nWhich letter belongs in the cell marked with '?'?"
        
        options = []
        letters = ['A', 'B', 'C', 'D', 'E']
        correct_id = ''
        for idx, letter in enumerate(letters):
            options.append({"id": chr(97 + idx), "text": letter})
            if letter == target_val:
                correct_id = chr(97 + idx)
                
        explanation = f"By applying the rules of Latin squares (no duplicate letters in any row or column), the cell marked '?' is uniquely forced to be {target_val}. You can deduce this by observing the row and column constraints that eliminate all other possibilities."
        
        # Escape single quotes for SQL
        prompt_sql = prompt.replace("'", "''")
        explanation_sql = explanation.replace("'", "''")
        options_sql = json.dumps(options).replace("'", "''")
        
        hash_hex = hashlib.md5(f"latin_batch2_{i}".encode()).hexdigest()
        q_id = f"{hash_hex[:8]}-{hash_hex[8:12]}-{hash_hex[12:16]}-{hash_hex[16:20]}-{hash_hex[20:32]}"
        
        val = f"('{q_id}', 'latin_square', '{prompt_sql}', '{options_sql}'::jsonb, '{correct_id}', '{explanation_sql}', '{diff}', 'aaaaaaaa-0000-0000-0000-000000000001', 75, 'standard_5x5', 'logical_deduction', 'latin_batch2_{i}')"
        values.append(val)
        
    sql_statements.append(",\n".join(values))
    sql_statements.append("\nON CONFLICT (id) DO NOTHING;\n")
    
    with open(output_file, "w", encoding="utf-8") as f:
        f.write("\n".join(sql_statements))
        
    print(f"Successfully generated {num_questions} Latin square questions in {output_file}")

if __name__ == "__main__":
    main()

File: scripts/generate-latin-squares.js
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

function uuid() {
  return crypto.randomUUID();
}

function randInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function shuffle(array) {
  const arr = [...array];
  for (let i = arr.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
  return arr;
}

// Generate base latin square
function generateBaseSquare(size) {
  const sq = [];
  for (let r = 0; r < size; r++) {
    const row = [];
    for (let c = 0; c < size; c++) {
      row.push((r + c) % size);
    }
    sq.push(row);
  }
  return sq;
}

// Shuffle rows, cols, and symbols
function randomizeSquare(sq, size) {
  let rs = [...Array(size).keys()];
  let cs = [...Array(size).keys()];
  let syms = [...Array(size).keys()];
  
  rs = shuffle(rs);
  cs = shuffle(cs);
  syms = shuffle(syms);
  
  const res = [];
  for (let r = 0; r < size; r++) {
    const row = [];
    for (let c = 0; c < size; c++) {
      row.push(syms[sq[rs[r]][cs[c]]]);
    }
    res.push(row);
  }
  return res;
}

function copyGrid(grid) {
  return grid.map(r => [...r]);
}

// Simple deduction solver
function solveGrid(grid, size) {
  let progress = true;
  const g = copyGrid(grid);
  
  while (progress) {
    progress = false;
    for (let r = 0; r < size; r++) {
      for (let c = 0; c < size; c++) {
        if (g[r][c] !== null) continue;
        
        const possible = new Set([...Array(size).keys()]);
        for (let i = 0; i < size; i++) {
          if (g[r][i] !== null) possible.delete(g[r][i]);
          if (g[i][c] !== null) possible.delete(g[i][c]);
        }
        
        if (possible.size === 1) {
          g[r][c] = Array.from(possible)[0];
          progress = true;
        }
      }
    }
  }
  
  return g;
}

function canSolveTarget(grid, size, targetR, targetC) {
  const solved = solveGrid(grid, size);
  return solved[targetR][targetC] !== null;
}

function generatePuzzle(size, difficulty) {
  const base = generateBaseSquare(size);
  const solution = randomizeSquare(base, size);
  
  // Choose random symbols
  const allSymbolSets = [
    ['A', 'B', 'C', 'D', 'E'],
    ['1', '2', '3', '4', '5'],
    ['+', '-', 'x', '/', '='],
    ['α', 'β', 'γ', 'δ', 'ε']
  ];
  const symbolSet = randChoice(allSymbolSets).slice(0, size);
  
  const targetR = randInt(0, size - 1);
  const targetC = randInt(0, size - 1);
  const targetVal = solution[targetR][targetC];
  
  // Start with full grid (except target)
  let puzzle = copyGrid(solution);
  puzzle[targetR][targetC] = null;
  
  // Create list of positions to try removing
  const pos = [];
  for (let r = 0; r < size; r++) {
    for (let c = 0; c < size; c++) {
      if (r !== targetR || c !== targetC) pos.push({r, c});
    }
  }
  const shuffledPos = shuffle(pos);
  
  let removedCount = 0;
  for (const p of shuffledPos) {
    const backup = puzzle[p.r][p.c];
    puzzle[p.r][p.c] = null;
    
    if (!canSolveTarget(puzzle, size, targetR, targetC)) {
      // Revert if it becomes unsolvable
      puzzle[p.r][p.c] = backup;
    } else {
      removedCount++;
    }
  }
  
  // Depending on difficulty, we might want to put some back to make it easier
  if (difficulty === 'easy') {
    // Put back ~30% of removed
    const toRestore = Math.floor(removedCount * 0.3);
    let restored = 0;
    for (const p of pos) {
      if (puzzle[p.r][p.c] === null && restored < toRestore) {
        puzzle[p.r][p.c] = solution[p.r][p.c];
        restored++;
      }
    }
  }
  
  // Format grid
  const finalGrid = [];
  for (let r = 0; r < size; r++) {
    const row = [];
    for (let c = 0; c < size; c++) {
      if (r === targetR && c === targetC) row.push('?');
      else if (puzzle[r][c] === null) row.push('');
      else row.push(symbolSet[puzzle[r][c]]);
    }
    finalGrid.push(row);
  }
  
  return {
    solutionGrid: solution.map(row => row.map(v => symbolSet[v])),
    finalGrid,
    symbolSet,
    targetR,
    targetC,
    targetVal: symbolSet[targetVal]
  };
}

function randChoice(arr) {
  return arr[randInt(0, arr.length - 1)];
}

function generateQuestion(difficulty) {
  const size = (difficulty === 'easy' || difficulty === 'medium') ? 4 : randChoice([4, 5]);
  const p = generatePuzzle(size, difficulty);
  
  const optionsRaw = p.symbolSet.map(s => ({
    id: s.toLowerCase(),
    text: s,
    isCorrect: s === p.targetVal
  }));
  
  // We need exactly 4 options. If size is 5, drop one incorrect option.
  let finalOpts = optionsRaw;
  if (size === 5) {
    const correctOpt = optionsRaw.find(o => o.isCorrect);
    const incorrectOpts = optionsRaw.filter(o => !o.isCorrect);
    finalOpts = [correctOpt, ...shuffle(incorrectOpts).slice(0, 3)];
  }
  
  const shuffled = shuffle(finalOpts);
  const optionIds = ['a', 'b', 'c', 'd'];
  let correctOptionId = 'a';

  const optionsFormatted = shuffled.map((o, i) => {
    if (o.isCorrect) correctOptionId = optionIds[i];
    return { id: optionIds[i], text: o.text };
  });

  const visualData = {
    type: "latin_square",
    size: size,
    symbols: p.symbolSet,
    grid: p.finalGrid,
    target_cell: { row: p.targetR, col: p.targetC }
  };
  
  const explanation = `In a Latin Square, each symbol must appear exactly once in each row and column. By deducing which symbols cannot go in the empty cells based on existing symbols in the same row or column, we can determine the target cell (?) must be ${p.targetVal}.`;

  return {
    id: uuid(),
    module: 'latin_square',
    prompt_text: 'Which symbol replaces the question mark?',
    options: optionsFormatted,
    correct_option_id: correctOptionId,
    explanation_text: explanation,
    difficulty: difficulty,
    source: 'procedural_gen_2026',
    time_estimate_seconds: 75,
    visual_data: visualData,
    subtopic: `${size}x${size} Grid`
  };
}

export function generateBatch(count, difficulty) {
  const qs = [];
  for (let i = 0; i < count; i++) {
    qs.push(generateQuestion(difficulty));
  }
  return qs;
}

if (process.argv[1] === __filename) {
  const sample = generateBatch(1, 'hard');
  console.log(JSON.stringify(sample[0], null, 2));
}
